aws_security_group output panic

### Terraform Version
```
$ terraform -v
Terraform v0.8.7
```

### Affected Resource(s)
- aws_security_group

### Terraform Configuration Files
```hcl
terraform {
  required_version = ">= 0.8.7"
}

provider "aws" {
  region  = "us-west-2"
}

resource "aws_vpc" "main" {
  cidr_block              = "10.0.0.0/16"
}

resource "aws_security_group" "default" {
  vpc_id                  = "${aws_vpc.main.id}"

  ingress {
      from_port       = 22
      to_port         = 22
      protocol        = "tcp"
      security_groups = ["${aws_security_group.bastion.id}"]
  }

  ingress {
      from_port   = 8
      to_port     = 0
      protocol    = "icmp"
      self        = true
  }
}

resource "aws_security_group" "bastion" {
    vpc_id      = "${aws_vpc.main.id}"

    ingress {
        from_port       = 22
        to_port         = 22
        protocol        = "tcp"
        cidr_blocks     = ["0.0.0.0/0"]
    }

    egress {
        from_port       = 0
        to_port         = 0
        protocol        = "-1"
        cidr_blocks     = ["0.0.0.0/0"]
    }
}

## These cause terraform to completely implode, but only when any of the rules reference another security group.
output "default_security_group_ingress"                   { value = ["${aws_security_group.default.ingress}"] }
output "default_security_group_egress"                    { value = ["${aws_security_group.default.egress}"] }

## These work fine for security groups using `self` or `cidr_blocks`
output "bastion_security_group_ingress"                   { value = ["${aws_security_group.bastion.ingress}"] }
output "bastion_security_group_egress"                    { value = ["${aws_security_group.bastion.egress}"] }
```
### Panic Output
<https://gist.github.com/ctbarrett/b424693b5dbb185bce2c4f16fb204b2f>

### Expected Behavior
Expected plan output:
```
+ aws_security_group.bastion
    description:                          "Managed by Terraform"
    egress.#:                             "1"
    egress.482069346.cidr_blocks.#:       "1"
    egress.482069346.cidr_blocks.0:       "0.0.0.0/0"
    egress.482069346.from_port:           "0"
    egress.482069346.prefix_list_ids.#:   "0"
    egress.482069346.protocol:            "-1"
    egress.482069346.security_groups.#:   "0"
    egress.482069346.self:                "false"
    egress.482069346.to_port:             "0"
    ingress.#:                            "1"
    ingress.2541437006.cidr_blocks.#:     "1"
    ingress.2541437006.cidr_blocks.0:     "0.0.0.0/0"
    ingress.2541437006.from_port:         "22"
    ingress.2541437006.protocol:          "tcp"
    ingress.2541437006.security_groups.#: "0"
    ingress.2541437006.self:              "false"
    ingress.2541437006.to_port:           "22"
    name:                                 "<computed>"
    owner_id:                             "<computed>"
    vpc_id:                               "${aws_vpc.main.id}"

+ aws_security_group.default
    description:                           "Managed by Terraform"
    egress.#:                              "<computed>"
    ingress.#:                             "2"
    ingress.1927567964.cidr_blocks.#:      "0"
    ingress.1927567964.from_port:          "8"
    ingress.1927567964.protocol:           "icmp"
    ingress.1927567964.security_groups.#:  "0"
    ingress.1927567964.self:               "true"
    ingress.1927567964.to_port:            "0"
    ingress.~3457728489.cidr_blocks.#:     "0"
    ingress.~3457728489.from_port:         "22"
    ingress.~3457728489.protocol:          "tcp"
    ingress.~3457728489.security_groups.#: "<computed>"
    ingress.~3457728489.self:              "false"
    ingress.~3457728489.to_port:           "22"
    name:                                  "<computed>"
    owner_id:                              "<computed>"
    vpc_id:                                "${aws_vpc.main.id}"

+ aws_vpc.main
    cidr_block:                "10.0.0.0/16"
    default_network_acl_id:    "<computed>"
    default_route_table_id:    "<computed>"
    default_security_group_id: "<computed>"
    dhcp_options_id:           "<computed>"
    enable_classiclink:        "<computed>"
    enable_dns_hostnames:      "<computed>"
    enable_dns_support:        "true"
    instance_tenancy:          "<computed>"
    main_route_table_id:       "<computed>"


Plan: 3 to add, 0 to change, 0 to destroy.
```

### Actual Behavior
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.


panic: strconv.Atoi: parsing "~3457728489": invalid syntax

goroutine 858 [running]:
github.com/hashicorp/terraform/flatmap.expandArray(0xc42034c7b0, 0xc4201d461d, 0x7, 0x9, 0xc42048e3c8, 0xc420344d01)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/flatmap/expand.go:69 +0x6f2
github.com/hashicorp/terraform/flatmap.Expand(0xc42034c7b0, 0xc4201d461d, 0x7, 0x2, 0x2)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/flatmap/expand.go:26 +0x3fc
github.com/hashicorp/terraform/terraform.(*Interpolater).interpolateComplexTypeAttribute(0xc420534f90, 0xc4201d461d, 0x7, 0xc42034c7b0, 0x6b273a0, 0xc420345000, 0x9, 0x0, 0xc420345100)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/interpolate.go:642 +0x206
github.com/hashicorp/terraform/terraform.(*Interpolater).computeResourceVariable(0xc420534f90, 0xc4203455d8, 0xc4203ece40, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/interpolate.go:446 +0xb6b
github.com/hashicorp/terraform/terraform.(*Interpolater).valueResourceVar(0xc420534f90, 0xc4203455d8, 0xc4201d4602, 0x22, 0xc4203ece40, 0xc42034cc90, 0xc4203e48c0, 0x0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/interpolate.go:248 +0x7c
github.com/hashicorp/terraform/terraform.(*Interpolater).Values(0xc420534f90, 0xc4203455d8, 0xc4201d46f0, 0x0, 0xc42040f630, 0x10909a5)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/interpolate.go:85 +0x5f1
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).Interpolate(0xc420162c30, 0xc4203eccc0, 0x0, 0x0, 0xc42030a400, 0x0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:285 +0xe7
github.com/hashicorp/terraform/terraform.(*EvalWriteOutput).Eval(0xc4202ef600, 0x6ab6840, 0xc420162c30, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval_output.go:48 +0x96
github.com/hashicorp/terraform/terraform.EvalRaw(0x6a94800, 0xc4202ef600, 0x6ab6840, 0xc420162c30, 0xc420072050, 0x2, 0xc42034cc30, 0x2b)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc4202ef640, 0x6ab6840, 0xc420162c30, 0x2, 0x2, 0x4a6136d, 0x4)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0x6c
github.com/hashicorp/terraform/terraform.EvalRaw(0x6a944c0, 0xc4202ef640, 0x6ab6840, 0xc420162c30, 0x2b, 0x0, 0x0, 0x2b)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc42034cab0, 0x6ab6840, 0xc420162c30, 0x2, 0x2, 0x4a6136d, 0x4)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/hashicorp/terraform/terraform.EvalRaw(0x6a94280, 0xc42034cab0, 0x6ab6840, 0xc420162c30, 0x422a0c0, 0xc42042e99c, 0x3fe3940, 0xc42042e9c0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x175
github.com/hashicorp/terraform/terraform.Eval(0x6a94280, 0xc42034cab0, 0x6ab6840, 0xc420162c30, 0xc42034cab0, 0x6a94280, 0xc42034cab0, 0x4921b00)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x45aa380, 0xc42035d400, 0x0, 0x0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/terraform/graph.go:301 +0xd4d
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc4201d6710, 0xc420468370, 0xc4201d6720, 0xc420535770, 0xc4201d6730, 0x45aa380, 0xc42035d400, 0xc420501ec0, 0xc4203d42a0)
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/dag/dag.go:246 +0x237
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
	/private/tmp/terraform-20170217-2671-8s7aqj/terraform-0.8.7/src/github.com/hashicorp/terraform/dag/dag.go:255 +0x6bd



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1.  Reference list of ingress/egress rules
2. `terraform plan`

### Important Factoids
After commenting the output, I was able to successfully apply the configuration. Successive plan/apply operations worked after that, with the output value uncommented; e.g. after adding an egress rule to the `sg-default`:

```
$ terraform apply
aws_vpc.main: Refreshing state... (ID: vpc-d8ed8cbf)
aws_security_group.bastion: Refreshing state... (ID: sg-2351115b)
aws_security_group.default: Refreshing state... (ID: sg-2751115f)
aws_security_group.default: Modifying...
  egress.#:                           "0" => "1"
  egress.482069346.cidr_blocks.#:     "0" => "1"
  egress.482069346.cidr_blocks.0:     "" => "0.0.0.0/0"
  egress.482069346.from_port:         "" => "0"
  egress.482069346.prefix_list_ids.#: "0" => "0"
  egress.482069346.protocol:          "" => "-1"
  egress.482069346.security_groups.#: "0" => "0"
  egress.482069346.self:              "" => "false"
  egress.482069346.to_port:           "" => "0"
aws_security_group.default: Modifications complete

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate

Outputs:

bastion_security_group_egress = [
    {
        cidr_blocks = [0.0.0.0/0],
        from_port = 0,
        prefix_list_ids = [],
        protocol = -1,
        security_groups = [],
        self = 0,
        to_port = 0
    }
]
bastion_security_group_ingress = [
    {
        cidr_blocks = [0.0.0.0/0],
        from_port = 22,
        protocol = tcp,
        security_groups = [],
        self = 0,
        to_port = 22
    }
]
default_security_group_egress = [
    {
        cidr_blocks = [0.0.0.0/0],
        from_port = 0,
        prefix_list_ids = [],
        protocol = -1,
        security_groups = [],
        self = 0,
        to_port = 0
    }
]
default_security_group_ingress = [
    {
        cidr_blocks = [],
        from_port = 22,
        protocol = tcp,
        security_groups = [sg-2351115b],
        self = 0,
        to_port = 22
    },
    {
        cidr_blocks = [],
        from_port = 8,
        protocol = icmp,
        security_groups = [],
        self = 1,
        to_port = 0
    }
]
```

Note this block from the first plan output, when the output value was commented:
```
    ingress.~3457728489.cidr_blocks.#:     "0"
    ingress.~3457728489.from_port:         "22"
    ingress.~3457728489.protocol:          "tcp"
    ingress.~3457728489.security_groups.#: "<computed>"
    ingress.~3457728489.self:              "false"
    ingress.~3457728489.to_port:           "22"
```

It looks like the internal index value for the resource is generated with a non-numeric character, which is causing the panic: `panic: strconv.Atoi: parsing "~3457728489": invalid syntax`

--- a/flatmap/expand.go
+++ b/flatmap/expand.go
@@ -57,6 +57,7 @@ func expandArray(m map[string]string, prefix string) []interface{} {
 	// regardless of value, and expand them in numeric order.
 	// See GH-11042 for more details.
 	keySet := map[int]bool{}
+	computed := map[string]bool{}
 	for k := range m {
 		if !strings.HasPrefix(k, prefix+".") {
 			continue
@@ -71,6 +72,12 @@ func expandArray(m map[string]string, prefix string) []interface{} {
 		// skip the count value
 		if key == "#" {
 			continue
+		}
+
+		// strip the computed flag if there is one
+		if strings.HasPrefix(key, "~") {
+			key = key[1:]
+			computed[key] = true
 		}
 
 		k, err := strconv.Atoi(key)
@@ -88,7 +95,11 @@ func expandArray(m map[string]string, prefix string) []interface{} {
 
 	result := make([]interface{}, num)
 	for i, key := range keysList {
-		result[i] = Expand(m, fmt.Sprintf("%s.%d", prefix, key))
+		keyString := strconv.Itoa(key)
+		if computed[keyString] {
+			keyString = "~" + keyString
+		}
+		result[i] = Expand(m, fmt.Sprintf("%s.%s", prefix, keyString))
 	}
 
 	return result

--- a/flatmap/expand_test.go
+++ b/flatmap/expand_test.go
@@ -122,6 +122,19 @@ func TestExpand(t *testing.T) {
 
 		{
 			Map: map[string]string{
+				"computed_set.#":       "1",
+				"computed_set.~1234.a": "a",
+				"computed_set.~1234.b": "b",
+				"computed_set.~1234.c": "c",
+			},
+			Key: "computed_set",
+			Output: []interface{}{
+				map[string]interface{}{"a": "a", "b": "b", "c": "c"},
+			},
+		},
+
+		{
+			Map: map[string]string{
 				"struct.#":         "1",
 				"struct.0.name":    "hello",
 				"struct.0.rules.#": hil.UnknownValue,

